Porting to Sky Shader Material

CavemanIke opened the pull request:
Fixes #33
Fixes #12 
Fixes #4 

-----
This PR removes the dependency on a physical sky dome for both atmosphere and clouds, and moves the shader code for atmospheric and cumulus cloud rendering to a dedicated SkyMaterial.gdshader that the Environment node uses directly. This also means that debanding is available.

Moon rendering is no longer dependent on a separate instanced scene, and instead uses an equirectangular texture that is set directly on the shader. Lunar phases are directly dependent on the position of the moon relative to the sun, with the shadow in the correct position.

CavemanIke commented:
After looking through the current open issues this may resolve the following:
#12 - Since this is a sky material now, fog rendering is not a problem anymore.
#17 - Xtarsia mentioned that the compatibility renderer will likely work much better with Godot handling the sky stuff internally.
#33 - The technique I implemented for the moon (thanks to Kelvin van Hoorne) means the moon is always lit on the correct side.

TokisanGames commented on addons/sky_3d/src/Sky3D.gd:
```diff
@@ -51,7 +51,7 @@ func set_sky_enabled(value: bool) -> void:
 	if not sky:
 		return
-	sky.sky_visible = value
+	#sky.sky_visible = value
```
Let's not leave lines randomly commented. Why is this commented out? Should it be removed or reenabled?

TokisanGames commented on addons/sky_3d/src/Skydome.gd:
```diff
@@ -471,62 +388,27 @@ func update_moon_coords() -> void:
-	var wll = ScatterLib.compute_wavelenghts_lambda(atm_wavelenghts)
+	var wll = ScatterLib.compute_wavelenghts_lambda(atm_wavelengths)
 	var wls = ScatterLib.compute_wavlenghts(wll)
 	var betaRay = ScatterLib.compute_beta_ray(wls)
```
Please add types for all of these. The styling you've been doing everywhere is great.

TokisanGames commented on addons/sky_3d/src/Skydome.gd:
```diff
@@ -5,30 +5,101 @@
+@export var ground_color: Color = Color(0.3, 0.3, 0.3, 1.0): set = set_ground_color
+@export var horizon_level: float = 0.0: set = set_horizon_level
+
+@export_group("Sun")
```
Given the large nature of this file, let's move Sun exported variables down into the sun group of functions, and so on.

TokisanGames commented on demo/Sky3DDemo.tscn:
```diff
@@ -1,60 +1,146 @@
+environment = SubResource("Environment_1g8sw")
+camera_attributes = SubResource("CameraAttributesPractical_d0ed4")
 script = ExtResource("2_cejat")
-current_time = 6.3
```
Let's keep the demo starting early in the morning, with the reflection probe enabled.

TokisanGames reviewed (changes requested):
Great work with this. Here's the first round of notes.

* Since this combines multiple shaders, the unused originals can be removed. cloudscumulus, pervertex, sky, simple moon and uid files.


* This doesn't ever turn off the lights. This means Godot is calculating two directional lights every frame. Previously it only ever did one. There used to be a performance hit running two lights. I need to test in a more complex scene. Godot rebuilds shaders every time the number of lights change, so if you're keeping the lights on so they illuminate under the horizon, there's a chance the sun and moon overlap and the total number of lights changes, causing a shader recompilation. We need to do more testing on 2 light performance and shader stutters if the lights are toggled when no longer used.
 
* I get these errors at startup, and some of the features in sky3d don't work likely because of it:
```
SCRIPT ERROR: Invalid call. Nonexistent function 'set_shader_parameter' in base 'PhysicalSkyMaterial'.
          at: Skydome.build_scene (res://addons/sky_3d/src/Skydome.gd:205)
SCRIPT ERROR: Invalid assignment of property or key 'sun_light_enable' with value of type 'bool' on a base object of type 'Node (Skydome)'.
          at: Sky3D.set_lights_enabled (res://addons/sky_3d/src/Sky3D.gd:65)
SCRIPT ERROR: Invalid assignment of property or key 'sun_light_enable' with value of type 'bool' on a base object of type 'Node (Skydome)'.
          at: Sky3D.set_lights_enabled (res://addons/sky_3d/src/Sky3D.gd:65)
```

* Sky3D/visibility/sky_enabled should disable the whole sky, eg going into a cave. This might mean storing the skyshader resource and removing it until reenabled. Setting energy to 0 is also fine unless there's a performance cost to keeping it running.
* Sky3D/visibility/lights_enabled doesn't turn off the lights, as it should.
* Sky3D/visibility/clouds_enabled works properly.
* Sky3d_enabled should disable all of those and time.
* This removed the screen space fog. I'd like to see it reenabled as another option.

* Is there something you can do to make this work with compatibility mode? See @Xtarsia 's change for the previous version. https://github.com/TokisanGames/Sky3D/pull/32/files

* Overall the PR is a little messy, as it includes changes to the demo scenes, and the upgrade to 4.4, as well as overhauling the skyshader. For an easier review and bug fixing later, I'd separate the latter from the former. This is fine now, but for the future aim for clean PRs.

* Finally, there is no rush on this. I have many things to review, so am not waiting for you to finish this today. When convenient is fine.

CavemanIke commented:
> * This doesn't ever turn off the lights. This means Godot is calculating two directional lights every frame. Previously it only ever did one. There used to be a performance hit running two lights. I need to test in a more complex scene. Godot rebuilds shaders every time the number of lights change, so if you're keeping the lights on so they illuminate under the horizon, there's a chance the sun and moon overlap and the total number of lights changes, causing a shader recompilation. We need to do more testing on 2 light performance and shader stutters if the lights are toggled when no longer used.

When toggling the lights on and off the shader will see only one light and default it to `LIGHT0`. This means that if the sun light were turned off, the moon light would become the new sun. Their energy is getting reduced to zero when they go below the horizon however, so they are not illuminating anything from underneath the map. I am uncertain of performance impacts for a light that is still "on" with its energy at zero.

CavemanIke commented:
Update:
- Removed all but AtmFog shader files. Work on re-implementing the screen space fog shader is still pending.
- Moved @export variables above associated functions.
- Removed commented-out code inadvertently left in place.
- Added missing static typing to variables.
- Updated Cloud Visibility to toggle uniform in shader and disables all cloud processing entirely.
- Updated Sky3D Enabled to toggle uniform in shader and disables all sky/clouds rendering and processing.
- Fixed Sky3D Enabled toggle not disabling lights.
- Fixed moon texture not loading on Sky3D instantiation.
- Reimplemented Reflected Energy to correspond with old PhysicalSkyMaterial behavior. Note that because there is no longer a sky dome this affects both the sky and object reflections, so the behavior will overall be different.
- Organized shader parameters into appropriate groups.

TokisanGames commented:
I've confirmed this fixes all of the issues added to Op. All three fogs work, and fixing the moon is fantastic. Very nice.

The Sky doesn't work on the compatibility renderer at all. 
The console dumps the sky shader and these errors:
```
ERROR: SkyShaderGLES3: Fragment shader compilation failed:
0(388) : error C1503: undefined variable "time"
0(572) : error C1503: undefined variable "time"
0(594) : error C1503: undefined variable "time"

   at: _display_error_with_code (drivers/gles3/shader_gles3.cpp:265)
ERROR: Method/function failed.
   at: _compile_specialization (drivers/gles3/shader_gles3.cpp:407)
```

I can't tell if the fog screenspace shader is working, though #32 is probably the fix for it.

Still doing a little more testing on it.

CavemanIke commented:
Fixes for the shader not working in compatibility and other notes:
- Added a `_time` uniform to the sky shader, replaced all instances of the `TIME` builtin, and set up the TimeOfDay script to update this property. This allows the shader to function in the Compatibility Renderer. This also has the effect of the clouds now pausing movement when time is disabled. Known issues:
  - Because this depends on TimeOfDay's update process which uses `Time.get_ticks_msec()`, clouds will appear to "jump" when editor or game time is reenabled if enough time has passed.
  - Cloud movement does not look as smooth on account of the default update interval. This can be remedied by setting the update interval a bit lower.
- Set the Sky's process mode to real time. Previously this was being automatically set to real time internally since the TIME builtin was being used. Without this, there is a very noticeable flickering that occurs on the ground and updates to the radiance cubemap causing artifacts.
- Fixed the Reflection Probe not culling the same layer as the lights. The ground is now properly lit over the course of the day and night. Can you check to see if this fixed the reflected energy issue you saw?

Future PR Notes:
- Fog screen space shader is not working yet in Compatibility Renderer. This will soon be in process as I want to try to pull its functionality into the sky shader so it's no longer dependent on a plane glued to the camera.
- I've been outlining an eventual overhaul for the sky rotation and sun/moon positions. This will require redoing a lot of the current implementation, but will ultimately make things look a lot more accurate, the sun will no longer appear to slide around the star map, and will address both the star alignment and the moon gimbal lock problem. The gimbal lock is because right now the directional lamp is making use of `Transform3D.looking_at()`.

Other Notes:
- The moon texture pinching is because of how the texture is being applied and is not mapped/aligned properly. Previously it relied on a 3D sphere that was being pasted into the sky dome, but now the texture is being handled entirely in the shader without any mesh geometry involved. I will need to address this in the shader and will have to introduce some parameters for aligning the texture map. Unfortunately there will always be some amount of pinching occurring at the poles when using an equirectangular texture. This can be mostly addressed by using a cubemap texture. I have plans to create an option for the user to be able to use whichever they prefer.

TokisanGames commented:
> Introduce a _process() function into the Skydome script to constantly update the moon's transform basis in the shader.

I'd rather not do that. I previously removed _process. Moon updates can occur from the update timer.

> Cloud movement does not look as smooth on account of the default update interval. This can be remedied by setting the update interval a bit lower

Yes, I see it. And this is less performant than before.

I removed _process from the original because the sun and other elements don't need to be updated at 200-500fps, especially on a 96 minute day. Changing to updating at 10fps saved a lot of gdscript processing.

Now I see the clouds need more regular updating, so this needs to calculate all elements at 60fps just to look normal, which is a waste. If it were written in C++ it would be less of a concern.

Let's do a hybrid approach. Revert the clouds to update by the GPU using TIME as before for vulkan. If we're in compatibility mode, update as you are now with gdscript driving the shader uniform.

In the shader you can determine compatibility mode like this:
```
uniform float _time = 0;
#if CURRENT_RENDERER == RENDERER_COMPATIBILITY
    #define time _time
#else
    #define time TIME
#endif
```

In GDScript you can use 
`String(ProjectSettings.get_setting_with_override("rendering/renderer/rendering_method")).contains("gl_compatibility");`

Check this at start up, which enables updating the _time uniform on the update interval.


> Unfortunately there will always be some amount of pinching occurring at the poles when using an equirectangular texture.

The slight warping of the moon texture at the edges of the sphere is fine and normal. What I showed is that the moon has rotated 10-20 degrees where the pinched pole is facing us. That should never happen. We should never see the pole.

![{746C0445-3346-4A58-9947-2334FC4F195B}](https://github.com/user-attachments/assets/34534c37-cd2b-4b9b-9f29-6859225b826c)

Manually adjusting azimuth to any value caused the projection to snap back into place with the light side properly facing us. 

I haven't determine what causes this, but I suspect it is an initialization issue based on this symptom and that there are a handful of other initialization issues.

I'm also troubleshooting a strange issue with the moon. In my game I have one scene with just the Sky3D tree, and it works fine like in the demo. The moon is normal sized and rotates with the sky. I have another scene that is nearly identical, but when I load it up, the moon is 100x bigger. If I switch scenes it returns to normal size, but when I change the time, the moon rotates without translating, like a spinning ball in the sky. It doesn't rotate with the sky. The scene files are virtually the same. I haven't tracked it down yet.

> Because this depends on TimeOfDay's update process which uses Time.get_ticks_msec(), clouds will appear to "jump" when editor or game time is reenabled if enough time has passed.

pause() and resume() are designed to avoid jumps. When resumed, no matter how long has passed, only one update_interval passes. I'm not concerned about enable_*_time jumps.

> Fixed the Reflection Probe not culling the same layer as the lights. The ground is now properly lit over the course of the day and night. Can you check to see if this fixed the reflected energy issue you saw?

I did notice an issue with the reflection probe not updating when first running the demo, but then it fixed itself. 🤷‍♂️ 

My current issue with reflected light has nothing to do with reflection probe. It is that the current value isn't being applied on startup, only if I change the value. Another init issue.


Fog and alignment issues let's discuss on those other pages.

CavemanIke commented:
Compatibility Renderer:
- `#if..#define` conditionals were added to the shader to define a `time` variable with either `_time` or `TIME` based on whether or not the engine is using compatibility renderer.
- Added a variable in TimeOfDay to check if in compatibility mode and therefore update the `_time` property. Since Compatibility relies on external updates from the Sky3D script, cloud movement will not be as smooth. Setting the update time interval a little lower may alleviate this at the cost of a wee bit of performance from the increased updates per frame.
  - The above is necessary due to an engine bug wherein the `TIME` builtin cannot be used in custom functions other than `sky()` in Compatibility mode.
  - This has the side effect of clouds pausing their movement if Game or Editor Time is disabled.
- Added Xtarsia's fix to get screen space fog working in the Compatibility renderer. Fog settings will need different adjustments compared to Forward+ to achieve the desired look.

Moon:
- Moon texture projection now updates when the date/time properties on TimeOfDay in the inspector are modified and will maintain its "rotation".
  - Caveat: If the rotation of the moon directional light is modified manually or through code for some reason this will result in the moon's transform matrix not being updated until `SkyDome.update_moon_coords()` is called and will cause the projection to be mapped/rotated incorrectly.
  - To-do: Introduce parameters so the texture alignment can be manually adjusted.

Star Map Alignment and Rotation:
- Introduced new properties into the shader and GDScript to allow for alignment of star maps. The provided default values work for most star maps that are in galactic coordinates format and should require little to no adjustment.
  - To-do: Rework the alignment to be more intuitive. Currently it is quite fiddly.
- Introduced properties to the shader and GDScript to directly control sky tilt and rotation for a more intuitive experience.
- Time of day now directly modifies the new sky rotation property and creates a more realistic sky simulation at any latitude. Latitude, Longitude, and UTC settings will also tilt and rotate the sky as expected.
- Star map texture is now properly displayed in the correct orientation (was previously mirrored).
- This deprecates the need for the deep space matrix calculations. Work is pending to properly remove those parts of code no longer needed for this purpose.
- This implementation is not very accurate yet. The sun and moon positions are still relatively incorrect in relation to the sky. This will be addressed in a future update or PR, but is still more accurate than previously and allows for greater customization and adjustments.

JohnC0de commented:
Hi @CavemanIke and @TokisanGames,

Amazing work on this PR! I’ve integrated the changes from this branch into my Godot 4.5 game, which requires compatibility mode, and I’m thrilled with how well it’s performing. The compatibility has been a game-changer for my project, and I haven’t noticed any major issues in my scenes so far.

Looking forward to seeing this merged—great job, and thanks for the hard work!

geowarin commented on addons/sky_3d/src/Skydome.gd:
```diff
@@ -1288,18 +1144,24 @@ func set_stars_scintillation_speed(value: float) -> void:
+func set_clouds_visible(value: bool) -> void:
+	if value == clouds_visible:
+		return
+	clouds_visible = value
```
```suggestion
	if !is_scene_built:
		return
	clouds_visible = value
```

TokisanGames commented:
Thanks for the fix and testing.

> I noticed that fps seems to be a bit worse than before but frame time seems much more stable.

Performance should be better moving from 5 shaders to 2. FPS should be inconsequential on modern cards and hard to measure sub 2ms.

TokisanGames commented:
This PR is getting long in the tooth. Let's not add any other new things to it, nor unrelated fixes, which should wait for new PRs. It makes it harder to review because now I have new things to test. I want to get it merged in soon. You can make a local branch off of this one if you want to commit unrelated things, and rebase it later.

* [x] The issues with TIME in the sky material should be reported upstream. However, we can do a better workaround. TIME does exist in compatibility mode, it just can't be accessed in subfunctions. You can grab it in the main function, then pass it as a parameter to all subfunctions. I've done this and reverted all of the CPU time code.

* [x] sky3d_enabled doesn't toggle fog_enabled

* [x] Reflection probe cull_mask
> Fixed the Reflection Probe not culling the same layer as the lights. The ground is now properly lit over the course of the day and night. Can you check to see if this fixed the reflected energy issue you saw?

The Demo reflection probe cull_mask is set only to layers 1/3. I guess it wasn't on 1, which you fixed. But I don't see anything else on 3, or any layer other than 1. I think we can just reset it to the default. This is unrelated to the reflection issues I was mentioning.

**Todo before merge:**

* [ ] 1. Process mode
> Set the Sky's process mode to real time. Previously this was being automatically set to real time internally since the TIME builtin was being used. Without this, there is a very noticeable flickering that occurs on the ground and updates to the radiance cubemap causing artifacts.

This was done only in the demo. If it's necessary, it needs to be done in Sky3D._initialize(). However I don't see any difference between the various modes. I looked at shadows of the sun and reflections of the sky on the metal balls, but don't see any flickering. I do see a dramatic effect on lighting in my game without a reflection probe, so I'd rather expose this rather than enforce a default, or do both.

Can you demonstrate or elaborate on the flickering?

I do have constant flickering in my foliage shadows that is very annoying, but that doesn't change with this setting. It seems DL.pancake_size <20 causes flicker in my grass, but something else causes other flickers, even if the sun isn't moving.

-----
* [ ] 2. If lights_enabled is off, it turns off the directional lights (correct), but changing the time doesn't move the sun/moon in the sky shader (not correct). It should disable the DLs, and maybe also disable rendering the sky/moon. But if it is going to render them, it should move them with the time change. I think rendering is fine. We would turn off the lights if we're going into a cave or building. 

Related, I also figured out that if the moon is visible and lights_enabled is off, when the time changes, the moon rotates in place, instead of rotating in the sky. This is what I was seeing in my game.

----
My remaining testing:
* Review all lighting parameters and defaults and ensure the current ones perform different operations.
* Not all settings are applied at load (eg reflected energy)
* Moon size is wonky/not working in OOTA, but looks ok in the Sky3D demo

Notes for the future, not for this PR:
* Compatibility mode Reflection Probe breaks? the reflections on the teapots. It looks strange. Can be fixed later.
* TOD replace get_property_list() w/ exports

TokisanGames commented:
Ok, reflected energy can go away. It essentially lights the same, but kills the fog. The others are unique. Done.

The second direction light, costs maybe 5 frames (out of 35) on my 3070. It also changes the appearance of the scene, even with energy at 0. The shadow surprisingly makes the scene slightly brighter. Disabling the shadow calculation, then toggling the light doesn't look any different.

SamDevelopsCode commented:
Tested this out to get volumetric fog in my current project working in which the fix is successful. Noticing a general jitter in the sky that does not happen with the previous master.

TokisanGames commented:
What specifically is jittering? Is it sensitive to camera movement? What's your update_interval?  Renderer? Versions? GPU? Can't do anything without details.

SamDevelopsCode commented:
> What specifically is jittering? Is it sensitive to camera movement? What's your update_interval? Renderer? Versions? Commit? GPU? Can't do anything without details.

Ah yea, forgot to provide anything useful. My bad.

1. Looks like the entire sky is jittering. The horizon is involved as well I believe.
2. Doesn't seem sensitive to camera movement.
3. Update interval is .016
4. Forward+
5. Version 4.4.1 Stable
6. Using sky3d version -> 2.1-dev
7. GPU = GTX3060Mobile (laptop)

Using Linux mint 

Heres a vid of the jitter

https://github.com/user-attachments/assets/7398796a-a4dd-46bd-b1bb-f31ccbed1252

TokisanGames commented:
I could finally look at this jitter on PC. I don't see this at all on my system, and I'm sure Cavemanike doesn't either or it would have come up.

You need to determine if it is the fog screen space shader, or the sky shader. If the latter, make a new scene, new environment, new sky, and explore the various native sky material options: procedural sky, physical sky, panorama sky with an HDR, and see which jitter if any.

I suspect the problem is either your driver, or a bug in Godot's sky material or rendering device that is triggered on your linux system. Or you enabled something in Godot that causes the jitter. Candidates would include physics interpolation, TAA, volumetric fog temporal reprojection, and possibly others. If it were in our control (eg our shader code), it would be jittering on all systems.

TokisanGames commented:
Linux mint nvidia display jitter
https://github.com/godotengine/godot/issues/104001

TokisanGames commented:
Nice work @CavemanIke thank you. We'll address pending issues in other tickets. From now on, let's do smaller, single issue PRs. They are easier to review and merge in. This one was an overhaul so was justifiably bigger.
